File: scripts/seed_turso_sample.mjs
// Sube ~50 filas sample del CSV a Turso. Más adelante subimos todo.
//
// Uso:
//   node scripts/seed_turso_sample.mjs <ruta_al_csv> [N=50]

import fs from "node:fs";
import readline from "node:readline";
import { createClient } from "@libsql/client";

const SCHEMA = `
CREATE TABLE IF NOT EXISTS contratos_csv (
  id_contrato         TEXT PRIMARY KEY,
  nombre_entidad      TEXT,
  nit_entidad         TEXT,
  departamento        TEXT,
  ciudad              TEXT,
  modalidad           TEXT,
  tipo_contrato       TEXT,
  estado_contrato     TEXT,
  es_pyme             TEXT,
  fecha_de_firma      TEXT,
  valor_del_contrato  REAL,
  proveedor           TEXT,
  documento_proveedor TEXT
);
`;

const cleanMoney = (s) => {
  if (s == null || s === "") return 0;
  const v = parseFloat(String(s).replace(/[$,\s]/g, ""));
  return Number.isFinite(v) ? v : 0;
};

// CSV parser muy simple — el SECOP usa comma + comillas dobles para quoting.
function parseCsvLine(line) {
  const out = [];
  let cur = "";
  let inQ = false;
  for (let i = 0; i < line.length; i++) {
    const c = line[i];
    if (inQ) {
      if (c === '"' && line[i + 1] === '"') {
        cur += '"';
        i++;
      } else if (c === '"') {
        inQ = false;
      } else cur += c;
    } else {
      if (c === '"') inQ = true;
      else if (c === ",") {
        out.push(cur);
        cur = "";
      } else cur += c;
    }
  }
  out.push(cur);
  return out;
}

async function main() {
  const csvPath = process.argv[2];
  const N = parseInt(process.argv[3] || "50", 10);
  if (!csvPath) {
    console.error("Uso: node scripts/seed_turso_sample.mjs <csv> [N]");
    process.exit(1);
  }

  const url = process.env.TURSO_DATABASE_URL;
  const authToken = process.env.TURSO_AUTH_TOKEN;
  if (!url || !authToken) {
    console.error("Faltan TURSO_DATABASE_URL / TURSO_AUTH_TOKEN");
    process.exit(1);
  }

  const db = createClient({ url, authToken });
  console.log("[1/3] Creando schema...");
  await db.execute(SCHEMA);

  console.log(`[2/3] Leyendo ${N} filas del CSV...`);
  const rl = readline.createInterface({
    input: fs.createReadStream(csvPath, { encoding: "utf-8" }),
    crlfDelay: Infinity,
  });
  let header = null;
  let rows = [];
  for await (const line of rl) {
    if (!header) {
      header = parseCsvLine(line);
      continue;
    }
    if (rows.length >= N) break;
    const cells = parseCsvLine(line);
    if (cells.length < header.length) continue;
    const r = Object.fromEntries(header.map((h, i) => [h, cells[i]]));
    rows.push(r);
  }
  rl.close();
  console.log(`      filas leídas: ${rows.length}`);

  console.log("[3/3] Insertando en Turso...");
  let n = 0;
  for (const r of rows) {
    try {
      await db.execute({
        sql: `INSERT OR REPLACE INTO contratos_csv (
          id_contrato, nombre_entidad, nit_entidad, departamento, ciudad,
          modalidad, tipo_contrato, estado_contrato, es_pyme,
          fecha_de_firma, valor_del_contrato, proveedor, documento_proveedor
        ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
        args: [
          r["ID Contrato"] || "",
          r["Nombre Entidad"] || "",
          r["Nit Entidad"] || "",
          r["Departamento"] || "",
          r["Ciudad"] || "",
          r["Modalidad de Contratacion"] || "",
          r["Tipo de Contrato"] || "",
          r["Estado Contrato"] || "",
          r["Es Pyme"] || "",
          r["Fecha de Firma"] || "",
          cleanMoney(r["Valor del Contrato"]),
          r["Proveedor Adjudicado"] || "",
          r["Documento Proveedor"] || "",
        ],
      });
      n++;
    } catch (e) {
      console.warn(`  fila ${n}: ${e.message}`);
    }
  }

  const t = await db.execute("SELECT count(*) AS n FROM contratos_csv");
  console.log(`      insertadas: ${n}`);
  console.log(`      total en tabla: ${t.rows[0].n}`);
  console.log("Listo.");
}

main().catch((e) => {
  console.error("ERROR:", e);
  process.exit(1);
});

File: scripts/seed_turso_sample.py
"""Sube 50 filas sample del CSV a Turso para que el dashboard tenga algo
mientras subimos el dataset completo.

Uso:
    python scripts/seed_turso_sample.py "ruta\\al\\csv"

Crea la tabla `contratos_csv` con las columnas mínimas usadas por el
endpoint /api/turso/contratos.
"""
from __future__ import annotations

import os
import sys
from pathlib import Path

import pandas as pd

try:
    import libsql_experimental as libsql
except ImportError:
    print("pip install libsql-experimental")
    sys.exit(1)


SCHEMA_SQL = """
CREATE TABLE IF NOT EXISTS contratos_csv (
    id_contrato         TEXT PRIMARY KEY,
    nombre_entidad      TEXT,
    nit_entidad         TEXT,
    departamento        TEXT,
    ciudad              TEXT,
    modalidad           TEXT,
    tipo_contrato       TEXT,
    estado_contrato     TEXT,
    es_pyme             TEXT,
    fecha_de_firma      TEXT,
    valor_del_contrato  REAL,
    proveedor           TEXT,
    documento_proveedor TEXT
);
"""


def clean_money(s) -> float:
    if pd.isna(s):
        return 0.0
    s = str(s).strip().replace("$", "").replace(",", "").replace(" ", "")
    try:
        return float(s)
    except ValueError:
        return 0.0


def main() -> None:
    if len(sys.argv) < 2:
        sys.exit("Uso: python scripts/seed_turso_sample.py <ruta_csv>")

    csv_path = Path(sys.argv[1]).resolve()
    if not csv_path.exists():
        sys.exit(f"No existe: {csv_path}")

    url = os.environ.get("TURSO_DATABASE_URL")
    auth_token = os.environ.get("TURSO_AUTH_TOKEN")
    if not url or not auth_token:
        sys.exit("Configura TURSO_DATABASE_URL y TURSO_AUTH_TOKEN en el entorno")

    print(f"[1/3] Leyendo CSV (primeras 50 filas)...")
    df = pd.read_csv(csv_path, low_memory=False, nrows=50)
    print(f"      filas: {len(df)} cols: {df.shape[1]}")

    print("[2/3] Conectando a Turso...")
    conn = libsql.connect("seed.db", sync_url=url, auth_token=auth_token)
    conn.execute(SCHEMA_SQL)

    print("[3/3] Insertando 50 filas sample...")
    insert_sql = """
        INSERT OR REPLACE INTO contratos_csv (
            id_contrato, nombre_entidad, nit_entidad, departamento, ciudad,
            modalidad, tipo_contrato, estado_contrato, es_pyme,
            fecha_de_firma, valor_del_contrato, proveedor, documento_proveedor
        ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
    """
    n = 0
    for _, r in df.iterrows():
        try:
            conn.execute(
                insert_sql,
                (
                    str(r.get("ID Contrato", "")),
                    str(r.get("Nombre Entidad", "")),
                    str(r.get("Nit Entidad", "")),
                    str(r.get("Departamento", "")),
                    str(r.get("Ciudad", "")),
                    str(r.get("Modalidad de Contratacion", "")),
                    str(r.get("Tipo de Contrato", "")),
                    str(r.get("Estado Contrato", "")),
                    str(r.get("Es Pyme", "")),
                    str(r.get("Fecha de Firma", "")),
                    clean_money(r.get("Valor del Contrato")),
                    str(r.get("Proveedor Adjudicado", "")),
                    str(r.get("Documento Proveedor", "")),
                ),
            )
            n += 1
        except Exception as e:
            print(f"  fila {n}: {e}")
    conn.commit()
    conn.sync()

    cur = conn.execute("SELECT count(*) FROM contratos_csv")
    print(f"      total en tabla tras insert: {cur.fetchone()[0]}")
    print("Listo.")


if __name__ == "__main__":
    main()
